File: 0036-valid-sudoku/0036-valid-sudoku.py
class Solution:
    def isValidSudoku(self, board: List[List[str]]) -> bool:
        rowSet = [set() for _ in range(9)]
        colSet = [set() for _ in range(9)]
        gridSet = [set() for _ in range(9)]

        for i in range(9):
            for j in range(9):
                gridNo = (i // 3) * 3 + (j // 3)

                if (board[i][j] != '.'):
                    isPresentInRow = board[i][j] in rowSet[i]
                    isPresentInCol = board[i][j] in colSet[j]
                    isPresentInGrid = board[i][j] in gridSet[gridNo]
                
                    if (isPresentInRow or isPresentInCol or isPresentInGrid):
                        return False

                    rowSet[i].add(board[i][j])
                    colSet[j].add(board[i][j])
                    gridSet[gridNo].add(board[i][j])
            
        return True
        

File: 0036-valid-sudoku/0036-valid-sudoku.cpp
class Solution {
public:
    bool isValidSudoku(vector<vector<char>>& board) {
        vector<unordered_set<char>> rowSet(9);
        vector<unordered_set<char>> colSet(9);
        vector<unordered_set<char>> gridSet(9);

        for (int i = 0; i < 9; i++) {
            for (int j = 0; j < 9; j++) {
                int gridNo = (i / 3) * 3 + (j / 3);

                if (board[i][j] != '.') {
                    bool isPresentInRow = rowSet[i].count(board[i][j]);
                    bool isPresentInCol = colSet[j].count(board[i][j]);
                    bool isPresentInGrid = gridSet[gridNo].count(board[i][j]);

                    if (isPresentInRow || isPresentInCol || isPresentInGrid) {
                        return false;
                    }

                    rowSet[i].insert(board[i][j]);
                    colSet[j].insert(board[i][j]);
                    gridSet[gridNo].insert(board[i][j]);
                }
            }
        }

        return true;
    }
};